### PR TITLE
Surface area review

### PR DESCRIPTION
**Pull request by L-Applin**
## Surface Area Review Changes
- Remove `SplitAsyncResponseTransformer`, replacing it with `AsyncRersponseTransformer.SplitResult` inner interface with default internal implementation, with builder.
- renamed `returnFuture` to `resultFuture`.
- check and guard against for negative buffer size, renamed to `bufferSizeInBytes`.
- add logic to not perform multipart if part number or range is specified.
- return response from first call to get-object.

## Other changes
- Small refactoring: moved some logic into `DownloadObjectHelper` to be consistent with upload/copy.

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java [hunk 1]
```diff
@@ -218,7 +222,6 @@ public CompletableFuture<ResponseT> prepare() {
         public void onResponse(ResponseT response) {
             if (onResponseCalled.compareAndSet(false, true)) {
                 log.trace(() -> "calling onResponse on the upstream transformer");
-                // todo: should we send back the first response to the upstreamResponseTransformer as-is?
```
Comment (anirudh9391) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java:221]: Did we decide to not do this ?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java [hunk 2]
```diff
@@ -308,7 +311,7 @@ public static <ResponseT, ResultT> Builder<ResponseT, ResultT> builder() {
 
     public static final class Builder<ResponseT, ResultT> {
 
-        private long maximumBufferSize;
+        private Long maximumBufferSize;
```
Comment (anirudh9391) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java:314]: If it is null, does it defer to default (32MB If i remember correctly ) ?

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java [hunk 3]
```diff
@@ -51,7 +49,7 @@ class MultipartDownloadIntegrationTest {
 
     private S3AsyncClient s3;
     private final SplittingTransformerConfiguration splitConfig = SplittingTransformerConfiguration.builder()
-                                                                                                   .bufferSize(1024 * 1024 * 32L)
+                                                                                                   .bufferSizeInBytes(1024 * 1024 * 32L)
```
Comment (anirudh9391) [inline services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java:52]: nit : Would improve readability if we had a variable denoting that this is the default value : 
```
private final int DEFAULT_BUFFERED_SIZE_IN_BYTES = 1024 * 1024 * 32L; 
```
With a comment that the default is 32MB.

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java [hunk 4]
```diff
@@ -63,24 +61,24 @@ void init() {
...
-        ResponseBytes<GetObjectResponse> res = split.preparedFuture().join();
+        ResponseBytes<GetObjectResponse> res = split.resultFuture().join();
         log.info(() -> "complete");
         byte[] bytes = res.asByteArray();
         log.info(() -> String.format("Byte len: %s", bytes.length));
         assertThat(bytes.length).isEqualTo(fileTestSize * 1024 * 1024);
     }
 
-    @Test
+    // @Test
```
Comment (anirudh9391) [inline services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java:81]: same as above

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java (hunk 1 above)
Comment (L-Applin) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java:221]: That question was part of the surface area review. In v2, the existing CRT client does return the first response, and in v1 the multipart is done all in transfer-manager so this behaviour does not apply, the GetObjectResponse is never returned to the user. Therefore, returning the first response seems appropriate.

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java [hunk 5]
```diff
@@ -63,24 +61,24 @@ void init() {
                                .build();
     }
 
-    @Test
+    // @Test
```
[the remark this answers is not in the document]
Comment (L-Applin) [reply services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java:64]: Intentional yes, but please ignore this class, it is only used manually by me to test real world scenario with s3 during development. But I'll add a todo to remove it before release.

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java (hunk 3 above)
Comment (L-Applin) [reply services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadIntegrationTest.java:52]: like mentioned in the other comment, please ignore this class

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java (hunk 2 above)
Comment (L-Applin) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java:314]: It's the `MultipartS3AsyncClient` will default to 32MB, but here if it is null, the constructor throws a NPE

**Review by anirudh9391 — APPROVED** [review]
Approving

**Pull request merged**